### ISSUE TITLE
The question about 'gen_result' and 'greedy_res'?

### ISSUE THREAD

**Issue by WERush** [ruotianluo/self-critical.pytorch#6]
Hi, ruotian,

I am confused why the 'gen_result' in the (line1)[https://github.com/ruotianluo/self-critical.pytorch/blob/master/train.py#L125] and 'greedy_res' in the (line2)[https://github.com/ruotianluo/self-critical.pytorch/blob/master/misc/rewards.py#L32] are different.
I find both of them run the same code, 'model.sample()'. 

Another issue is that I run the code 'greedy_res, _ = model.sample(Variable(fc_feats.data, volatile=True), Variable(att_feats.data, volatile=True))' twice, and I get two different outputs. Can you explain that? Thanks very much.

**Comment by ruotianluo** [comment]
1. Difference is at sample_max
2. Because the model is in training mode, so the dropout will produce different results.

**Comment by WERush** [comment]
In my opinion, the 'greedy_res' should be extracted in the inference. Maybe it is better to turn off the training when computing the 'greedy_res'. Isn't it?

**Comment by ruotianluo** [comment]
It may. You can potentially try, see how the results differ.

**Comment by WERush** [comment]
Since I find the original paper said that "the reward obtained by the estimated sentence under the test-time inference procedure(greedy inference depicted)" in the Figure1.

**Comment by WERush** [comment]
OK, I will try.
